tests/test_browser.go

```diff
@@ -0,0 +1,297 @@
+		enableHTTPMultiBin = false
+		enableFileServer   = false
+	)
+	for _, opt := range opts {
+		switch opt := opt.(type) {
+		case withLaunchOptions:
+			launchOpts = opt
+		case httpServerOption:
+			enableHTTPMultiBin = true
+		case fileServerOption:
+			enableFileServer = true
+			enableHTTPMultiBin = true
+		case withContext:
+			ctx = opt
+		}
+	}
+
+	// create a k6 state
+	root, err := k6lib.NewGroup("", nil)
+	require.NoError(t, err)
+
+	samples := make(chan k6stats.SampleContainer, 1000)
+
+	state := &k6lib.State{
+		Options: k6lib.Options{
+			MaxRedirects: null.IntFrom(10),
+			UserAgent:    null.StringFrom("TestUserAgent"),
+			Throw:        null.BoolFrom(true),
+			SystemTags:   &k6stats.DefaultSystemTagSet,
+			Batch:        null.IntFrom(20),
+			BatchPerHost: null.IntFrom(20),
+			// HTTPDebug:    null.StringFrom("full"),
+		},
+		Logger:         logrus.StandardLogger(),
+		Group:          root,
+		BPool:          bpool.NewBufferPool(1),
+		Samples:        samples,
+		Tags:           k6lib.NewTagMap(map[string]string{"group": root.Path}),
+		BuiltinMetrics: k6metrics.RegisterBuiltinMetrics(k6metrics.NewRegistry()),
+	}
+
+	// configure the goja VM (1)
+	rt := goja.New()
+
+	// enable the HTTP test server only when necessary
+	var testServer *k6test.HTTPMultiBin
+	if enableHTTPMultiBin {
+		testServer = k6test.NewHTTPMultiBin(t)
+		state.TLSConfig = testServer.TLSClientConfig
+		state.Transport = testServer.HTTPTransport
+
+		err = rt.Set("http", k6common.Bind(rt, new(k6http.GlobalHTTP).NewModuleInstancePerVU(), &ctx))
+		require.NoError(t, err)
+	}
+
+	// configure the goja VM (2)
+	ctx = k6lib.WithState(ctx, state)
+	ctx = k6common.WithRuntime(ctx, rt)
+	rt.SetFieldNameMapper(k6common.FieldNameMapper{})
+
+	// launch the browser
+	bt := chromium.NewBrowserType(ctx).(*chromium.BrowserType)
+	b := bt.Launch(rt.ToValue(launchOpts))
+	t.Cleanup(func() {
+		select {
+		case <-ctx.Done():
+		default:
+			b.Close()
+		}
+	})
+
+	tb := &browser{
+		ctx:     bt.Ctx, // This context has the additional wrapping of common.WithLaunchOptions
+		rt:      rt,
+		state:   state,
+		http:    testServer,
+		samples: samples,
+		Browser: b,
+	}
+	if enableFileServer {
+		tb = tb.withFileServer()
+	}
+	return tb
+}
+
+// withHandle adds the given handler to the HTTP test server and makes it
+// accessible with the given pattern.
+func (b *browser) withHandle(pattern string, handler http.HandlerFunc) *browser {
+	if b.http == nil {
+		panic("You should enable HTTP test server, see: withHTTPServer option")
+	}
+	b.http.Mux.Handle(pattern, handler)
+	return b
+}
+
+const testBrowserStaticDir = "static"
+
+// withFileServer serves a file server using the HTTP test server that is
+// accessible via `testBrowserStaticDir` prefix.
+//
+// This method is for enabling the static file server after starting a test
+// browser. For early starting the file server see withFileServer function.
+func (b *browser) withFileServer() *browser {
```

Comment:
This is a method:

```golang
func (b *browser) withFileServer() *browser
```

This is a function:
```golang
func withFileServer() fileServerOption
```

Reply:
Ah OK, I missed that it's also a standalone function. Since methods in Go are just functions with a receiver, having both with the same name is slightly confusing, but not a big deal.

Reply:
Think of it like function overloading :P